### PR TITLE
fix: Calculate transaction fees in order

### PR DESCRIPTION
**Pull request by keroro520**
fix: The loading tx_fees is unordered HashSet [L150-L157](https://github.com/nervosnetwork/ckb/blob/develop/util/reward-calculator/src/lib.rs#L150-L157). But we use it as an ordered vector when calculates proposal_fees [L211]( https://github.com/nervosnetwork/ckb/blob/develop/util/reward-calculator/src/lib.rs#L211).

### CODE REVIEW

File: util/reward-calculator/src/lib.rs [hunk 1]
```diff
@@ -147,7 +147,7 @@ impl<'a, CS: ChainStore<'a>> RewardCalculator<'a, CS> {
         let mut proposed: HashSet<ProposalShortId> = HashSet::default();
         let mut index = parent.to_owned();
 
-        let committed_idx_proc = |hash: &H256| -> HashSet<ProposalShortId> {
+        let committed_idx_proc = |hash: &H256| -> Vec<ProposalShortId> {
```
Comment (doitian) [inline util/reward-calculator/src/lib.rs]: It's better to add a comment that it must return ids in the same order as the fees returned from `txs_fees_proc`.

**Comment by doitian** [comment]
Need to backport to v0.18.0?

**Comment by quake** [comment]
> Need to backport to v0.18.0?

Yes, need backport

**Comment by quake** [comment]
I'm confused about the usage of `Hashset#intersection` in the original code, we can use simple code since the intersection is checked by `target_proposals.remove(&id)` already:

```rust
        let committed_idx_proc = |hash: &H256| -> Vec<ProposalShortId> {
            store
                .get_block_txs_hashes(hash)
                .iter()
                .skip(1)
                .map(ProposalShortId::from_tx_hash)
                .collect()
        };

        let txs_fees_proc = |hash: &H256| -> Vec<Capacity> {
            store
                .get_block_ext(hash)
                .map(|block_ext| block_ext.txs_fees)
                .unwrap_or_default()
        };

        let committed_idx = committed_idx_proc(index.hash());

        for (id, tx_fee) in committed_idx
            .into_iter()
            .zip(txs_fees_proc(index.hash()).iter())
        {
            // target block is the earliest block with effective proposals for the parent block
            if target_proposals.remove(&id) {
                reward = reward.safe_add(tx_fee.safe_mul_ratio(proposer_ratio)?)?;
            }
        }
```

**Comment by keroro520** [comment]
> I'm confused about the usage of `Hashset#intersection` in the original code, we can use simple code since the intersection is checked by `target_proposals.remove(&id)` already:
> 
> ```rust
>         let committed_idx_proc = |hash: &H256| -> Vec<ProposalShortId> {
>             store
>                 .get_block_txs_hashes(hash)
>                 .iter()
>                 .skip(1)
>                 .map(ProposalShortId::from_tx_hash)
>                 .collect()
>         };
> 
>         let txs_fees_proc = |hash: &H256| -> Vec<Capacity> {
>             store
>                 .get_block_ext(hash)
>                 .map(|block_ext| block_ext.txs_fees)
>                 .unwrap_or_default()
>         };
> 
>         let committed_idx = committed_idx_proc(index.hash());
> 
>         for (id, tx_fee) in committed_idx
>             .into_iter()
>             .zip(txs_fees_proc(index.hash()).iter())
>         {
>             // target block is the earliest block with effective proposals for the parent block
>             if target_proposals.remove(&id) {
>                 reward = reward.safe_add(tx_fee.safe_mul_ratio(proposer_ratio)?)?;
>             }
>         }
> ```

I think that is a performance optimization. If there is not any intersection between `target_proposals` and `committed_idx`, then we can skip the for-loop below.

**Comment by u2** [comment]
Yes, it's a performance optimization. Try to reduce DB operation of `txs_fees_proc`

**Comment by quake** [comment]
> Yes, it's a performance optimization. Try to reduce DB operation of `txs_fees_proc`.

Get, we can use `filter_map` to resolve this issue, which avoids unnecessary `clone` and hashset comparison:

```rust
        let committed_idx: Vec<_> = store
            .get_block_txs_hashes(index.hash())
            .iter()
            .skip(1)
            .enumerate()
            .filter_map(|(i, hash)| {
                if target_proposals.remove(&ProposalShortId::from_tx_hash(hash)) {
                    Some(i)
                } else {
                    None
                }
            })
            .collect();

        if !committed_idx.is_empty() {
            let txs_fees = txs_fees_proc(index.hash());
            for i in committed_idx.into_iter() {
                reward = reward.safe_add(txs_fees[i].safe_mul_ratio(proposer_ratio)?)?;
            }
        }
```

**Comment by doitian** [comment]
The backport to rc/v0.18 failed:

test bot

**Pull request merged**